internal/util/telemetry/reporter_test.go

```diff
@@ -97,72 +97,106 @@ func TestNewReporterLock(t *testing.T) {
-			}
-
-			r, err := NewReporter(&opts)
-			assert.NoError(t, err)
-
-			// check initial state of provider, it has not called telemetry yet,
-			// no update is available and unaware of the latest version.
-			s := r.P.Get()
-			require.False(t, s.UpdateAvailable())
-			require.Equal(t, "", s.LatestVersion)
-
-			// call report to update the state of provider from telemetry.
-			r.report(testutil.Ctx(t))
-
-			// get updated the state of provider.
-			s = r.P.Get()
-			require.Equal(t, tc.updateAvailable, s.UpdateAvailable())
-			require.Equal(t, tc.latestVersion, s.LatestVersion)
-		})
+		LatestVersion:   "0.3.4",
+		UpdateAvailable: true,
 	}
+
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if r.Method == http.MethodPost {
+			w.WriteHeader(http.StatusCreated)
+			require.NoError(t, json.NewEncoder(w).Encode(telemetryResponse))
+		}
+	}))
+	defer ts.Close()
+
+	provider, err := state.NewProvider("")
+	require.NoError(t, err)
+
+	// Subtests are not run in parallel because they share the same telemetry mock.
+
+	t.Run("TelemetryEnabled", func(t *testing.T) {
+		opts := NewReporterOpts{
+			URL:           ts.URL,
+			F:             &Flag{v: pointer.ToBool(true)},
+			ConnMetrics:   connmetrics.NewListenerMetrics().ConnMetrics,
+			P:             provider,
+			L:             zap.L(),
+			ReportTimeout: 1 * time.Minute,
+		}
+
+		r, err := NewReporter(&opts)
+		require.NoError(t, err)
+
+		// Check the initial state of the provider, it has not called telemetry yet,
+		// no update is available and unaware of the latest version.
+		s := r.P.Get()
+		assert.False(t, s.IsUpdateAvailable)
+		assert.Equal(t, "", s.LatestVersion)
+
+		// Call the telemetry server and check the state of the provider to be updated.
+		r.report(testutil.Ctx(t))
+		s = r.P.Get()
+		assert.True(t, s.IsUpdateAvailable)
+		assert.Equal(t, "0.3.4", s.LatestVersion)
+
+		// Set update available to false on the telemetry side, and call the telemetry server again.
+		telemetryResponse.UpdateAvailable = false
+		r.report(testutil.Ctx(t))
+
+		// Expect the state of provider to be updated.
+		s = r.P.Get()
+		assert.False(t, s.IsUpdateAvailable)
+		assert.Equal(t, "0.3.4", s.LatestVersion)
+
+		// Set update available to true and update version, and call the telemetry server again.
+		//	mx.Lock()
+		telemetryResponse.UpdateAvailable = true
+		telemetryResponse.LatestVersion = "0.4.0"
+		r.report(testutil.Ctx(t))
+
+		// Expect the state and the version to be updated.
+		s = r.P.Get()
+		assert.True(t, s.IsUpdateAvailable)
+		assert.Equal(t, "0.4.0", s.LatestVersion)
+
+		// Disable telemetry and call the telemetry server again.
+		require.NoError(t, provider.Update(func(s *state.State) { s.Telemetry = pointer.ToBool(false) }))
+		r.report(testutil.Ctx(t))
+
+		// Expect no update to be available, and the latest version to be empty.
+		s = r.P.Get()
+		assert.False(t, s.IsUpdateAvailable)
+		assert.Equal(t, "", s.LatestVersion)
```

Comment:
@AlekSi please let me know if I understood this correctly: if telemetry is disabled, we reset the info about update availability and the latest version.

Reply:
There is no need to reset the info about the latest version anymore. We did that when there was only that field and packed two things into it. Now there is a separate field for that.

Reply:
fixed